### PR TITLE
15' + 1 hour record: Document-level shuffling

### PR DESCRIPTION
**Pull request by samacqua**
Instead of shuffling the order of a fixed set of batches every epoch, shuffle the order of the documents in the training set, then tokenize into batches + re-shuffle.

[15' log file](https://github.com/user-attachments/files/26917051/terminal.log):
Old record: 3.345
New record: 3.332

[1hr log file](https://github.com/user-attachments/files/26917044/terminal.log):
Old record: 3.211
New record: 3.204

This is taking one of the many components of #67 so that comparison to previous results is straightforward. This does not change the eval tokens or batching at all -- all it does is make the training shuffle the documents before re-batching. This adds more training diversity as any given token is preceded by different tokens every epoch.

I also added 2 hour record which was pre-empted while running. Submitting now, hoping someone can run it. The code is backwards compatible with `--no-doc-shuffle`, and I added some QOL things like saving the log, training file, and checkpoints to a file. This not only makes dev easier, but also replication. I hope people adopt this standard of attaching logs.

### CODE REVIEW

File: prepare_data.py [hunk 1]
```diff
@@ -15,176 +15,100 @@
...
-    write_datafile(val_path, val_sequences, BATCH_SIZE)
-    write_datafile(train_path, train_sequences, BATCH_SIZE)
+    write_datafile(val_path, val_tokens_arr, val_doc_starts, bos_id, VAL_SHUFFLE_SEED)
+    write_datafile(train_path, train_tokens_arr, train_doc_starts, bos_id, TRAIN_SHUFFLE_SEED)
 
-    # Verify hashes
     print()
-    verify_hash(val_path)
-    verify_hash(train_path)
+    for p in (val_path, train_path):
```
Comment (akshayvegesna) [inline prepare_data.py]: Can you add an assert that the hash doesn't change? It's nice to have confidence that nothing has changed upstream in the huggingface dataset.

File: two_hour/train.py [hunk 2]
```diff
@@ -12,12 +12,17 @@
 import math
```
Comment (akshayvegesna) [inline two_hour/train.py:12]: Since we don't have a record here, let's leave this file changes out until we do.

File: train.py [hunk 3]
```diff
@@ -106,6 +114,8 @@
 TOTAL_BATCH_SIZE = args.total_batch_size
 EVAL_TOKENS = 10_000_000
 DATA_DIR = "fineweb_data"
+BOS_ID = 50256  # <|endoftext|>
+RUNS_DIR = "runs"
```
Comment (akshayvegesna) [inline train.py:116]: Let's add runs/ to .gitignore

File: train.py [hunk 4]
```diff
@@ -145,6 +155,27 @@ def __init__(self): self.summary = {}
...
+        for stream in self.streams: stream.flush()
+    def isatty(self):
+        return any(getattr(stream, "isatty", lambda: False)() for stream in self.streams)
+    def fileno(self):
+        return self.streams[0].fileno()
+
+def resolve_run_dir(run_name):
+    if run_name:
+        return run_name, os.path.join(RUNS_DIR, run_name)
+    name = "".join(secrets.choice(string.ascii_lowercase + string.digits) for _ in range(6))
```
Comment (akshayvegesna) [inline train.py]: nit: I suggest we keep the datetime run id.

**Review by akshayvegesna — COMMENTED** [review]
Thank you, left a couple minor comments. To merge the update to prepare_data, we also need to ensure that unlimited/train.py (and any other code) doesn't break. Can you update the Dataloader with your no-document-level shuffling version that is equivalent to before so it works?

**Comment by samacqua** [comment]
Alright all have been changed. All the non-record dataloaders (unlimited/, two_hour/, research/) should be token-for-token identical to before. Thanks!

**Comment by akshayvegesna** [comment]
Thank you, looks good. Merging.
